WRPAccessControl.go

```diff
@@ -0,0 +1,158 @@
+
+// wrpPartnersAuthority defines the access policy for which WRP messages
+// are authorized against the partners credentials of the message creator
+type wrpPartnersAccess struct {
+	strict                  bool
+	receivedWRPMessageCount metrics.Counter
+}
+
+func (p *wrpPartnersAccess) withFailure(labelValues ...string) metrics.Counter {
+	if !p.strict {
+		return p.withSuccess(labelValues...)
+	}
+	return p.receivedWRPMessageCount.With(append(labelValues, OutcomeLabel, Rejected)...)
+}
+
+func (p *wrpPartnersAccess) withSuccess(labelValues ...string) metrics.Counter {
+	return p.receivedWRPMessageCount.With(append(labelValues, OutcomeLabel, Accepted)...)
+}
+
+//authorizeWRP runs the partners access policy against the WRP and returns an error if the check fails.
+//When the policy is not strictly enforced,
+// Additionally, when the policy is not a boolean is returned for failure cases where the policy autocorrects the WRP contents
+func (p *wrpPartnersAccess) authorizeWRP(ctx context.Context, message *wrp.Message) (bool, error) {
+	var (
+		auth, ok    = bascule.FromContext(ctx)
+		satClientID = "none"
+	)
+
+	if !ok {
+		p.withFailure(ClientIDLabel, satClientID, ReasonLabel, TokenMissing).Add(1)
+
+		if p.strict {
+			return false, ErrTokenMissing
+		}
+		return false, nil
+	}
+
+	token := auth.Token
+
+	if token.Type() != "jwt" {
+		p.withFailure(ClientIDLabel, satClientID, ReasonLabel, TokenTypeMismatch).Add(1)
+
+		if p.strict {
+			return false, ErrTokenTypeMismatch
+		}
+		return false, nil
+	}
+
+	attributes := token.Attributes()
+
+	if principal := token.Principal(); len(principal) > 0 {
+		satClientID = principal
+	}
+
+	allowedPartners, ok := attributes.GetStringSlice(basculechecks.PartnerKey)
+
+	if !ok || len(allowedPartners) < 1 {
+		p.withFailure(ClientIDLabel, satClientID, ReasonLabel, JWTPIDInvalid).Add(1)
+
+		if p.strict {
+			return false, ErrInvalidAllowedPartners
+		}
+
+		return false, nil
+	}
+
+	if len(message.PartnerIDs) < 1 {
+		p.withFailure(ClientIDLabel, satClientID, ReasonLabel, WRPPIDMissing).Add(1)
+
+		if p.strict {
+			return false, ErrPIDMissing
+		}
+
+		message.PartnerIDs = allowedPartners
+		return true, nil
+	}
+
+	if contains(allowedPartners, "*") {
+		p.withSuccess(ClientIDLabel, satClientID, ReasonLabel, JWTPIDWildcard).Add(1)
+		return false, nil
+	}
+
+	if isSubset(message.PartnerIDs, allowedPartners) {
+		p.withSuccess(ClientIDLabel, satClientID, ReasonLabel, WRPPIDMatch).Add(1)
+		return false, nil
+	}
+
+	p.withFailure(ClientIDLabel, satClientID, ReasonLabel, WRPPIDMismatch).Add(1)
+	if p.strict {
+		return false, ErrPIDMismatch
+	}
+
+	message.PartnerIDs = allowedPartners
+	return true, nil
+}
+
+//returns true if list contains str
+func contains(list []string, str string) bool {
+	for _, e := range list {
+		if e == str {
+			return true
+		}
+	}
+	return false
+}
+
+//returns true if a is a subset of b
+func isSubset(a, b []string) bool {
+	m := make(map[string]struct{})
```

Comment:
The idiomatic Set in golang is `map[string]bool`.  Among other things, this allows you to use defaulting to determine if something is in the set.  You don't have to use the two-argument form for lookups:

`
var set map[string]bool
set["included"] = true

if set["included"] { fmt.Println("this will execute") }
if set["nosuch"] { fmt.Println("this will not execute") }
`